### ISSUE TITLE
Puppet parser validate does not show error indicator

### ISSUE THREAD

**Issue by run2cmd** [dense-analysis/ale#2656]
## Information

VIM - Vi IMproved 8.1 (2018 May 18, compiled Jul 15 2019 20:54:41)
MS-Windows 64-bit GUI/console version with OLE support

Operating System: Windows 10

Puppet 5.5.6

## What went wrong
Looks like ALE doest not take into account `puppet parser validate` Error messages.

Puppet code example:
```
class test::class {
  include ::test::params
#}
```

The end of line is commented out and ALEInfo is reporting 
```
<<<OUTPUT STARTS>>>                                                                                                     Error:                                                                                                    Error: Could not parse for environment production: Syntax error at end of input (file: C:/Users/pbugala/AppData/Local/Temp/VINB0DF.tmp/test.pp) 
```

However I do not get the indicator in VIM that there is erorr. Also Status line is not reporting that
![image](https://user-images.githubusercontent.com/9575150/61784692-7063b780-ae0a-11e9-9008-243af0db3f3a.png)

I use custom linter status line
```
" Custom Ale Linter in status line
function! LinterStatus() abort
    let l:counts = ale#statusline#Count(bufnr(''))
    let l:all_errors = l:counts.error + l:counts.style_error
    let l:all_non_errors = l:counts.total - l:all_errors
    return l:counts.total == 0 ? 'OK' : printf(
    \ '%dW %dE',
    \ all_non_errors,
    \ all_errors
    \)
endfunction
```

### :ALEInfo

```

 Current Filetype: puppet
Available Linters: ['languageserver', 'puppet', 'puppetlint']
  Enabled Linters: ['languageserver', 'puppet', 'puppetlint']
 Suggested Fixers: 
  'puppetlint' - Run puppet-lint -f on a file.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
 Linter Variables:

let g:ale_puppet_languageserver_executable = 'puppet-languageserver'
let g:ale_puppet_puppet_executable = 'puppet'
let g:ale_puppet_puppet_options = '--render-as console'
let g:ale_puppet_puppetlint_executable = 'puppet-lint'
let g:ale_puppet_puppetlint_options = '--no-autoloader_layout-check'
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = 50
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '[%linter%][%severity%][%code%] %s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'puppet': ['puppetlint', 'trim_whitespace', 'remove_trailing_lines'], 'markdown': ['prettier', 'trim_whitespace', 'remove_trailing_lines'], 'ruby': ['rubocop', 'trim_whitespace', 'remove_trailing_lines'], 'yaml': ['prettier', 'trim_whitespace', 'remove_trailing_lines'], 'python': ['autopep8', 'isort', 'add_blank_lines_for_python_control_statements', 'trim_whitespace', 'remove_trailing_lines']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 0
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_linters_explicit = 0
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '[%linter%][%severity%][%code%] %s'
let g:ale_lsp_root = {}
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - failure) puppet-languageserver
(finished - exit code 0) 'cmd /s/c "puppet parser validate --color=false  --render-as console C:\Users\pbugala\AppData\Local\Temp\VIL8B54.tmp\test.pp"'

<<<NO OUTPUT RETURNED>>>

(finished - exit code 0) 'cmd /s/c "puppet-lint --no-autoloader_layout-check --log-format "-:%{line}:%{column}: %{kind}: [%{check}] %{message}" C:\Users\pbugala\AppData\Local\Temp\VIM8B55.tmp\test.pp"'

<<<NO OUTPUT RETURNED>>>

(executable check - failure) puppet-languageserver
(executable check - failure) puppet-languageserver
(finished - exit code 1) 'cmd /s/c "puppet parser validate --color=false  --render-as console C:\Users\pbugala\AppData\Local\Temp\VINB0DF.tmp\test.pp"'

<<<OUTPUT STARTS>>>
Error: Could not parse for environment production: Syntax error at end of input (file: C:/Users/pbugala/AppData/Local/Temp/VINB0DF.tmp/test.pp)
<<<OUTPUT ENDS>>>

(finished - exit code 0) 'cmd /s/c "puppet-lint --no-autoloader_layout-check --log-format "-:%{line}:%{column}: %{kind}: [%{check}] %{message}" C:\Users\pbugala\AppData\Local\Temp\VIPB0E0.tmp\test.pp"'

<<<OUTPUT STARTS>>>
-:2:11: warning: [relative_classname_inclusion] class included by relative name
<<<OUTPUT ENDS>>>

(executable check - failure) puppet-languageserver
(executable check - failure) puppet-languageserver
(finished - exit code 1) 'cmd /s/c "puppet parser validate --color=false  --render-as console C:\Users\pbugala\AppData\Local\Temp\VIQFCD9.tmp\test.pp"'

<<<OUTPUT STARTS>>>
Error: Could not parse for environment production: Syntax error at end of input (file: C:/Users/pbugala/AppData/Local/Temp/VIQFCD9.tmp/test.pp)
<<<OUTPUT ENDS>>>

(finished - exit code 0) 'cmd /s/c "puppet-lint --no-autoloader_layout-check --log-format "-:%{line}:%{column}: %{kind}: [%{check}] %{message}" C:\Users\pbugala\AppData\Local\Temp\VIRFCEA.tmp\test.pp"'

<<<NO OUTPUT RETURNED>>>
```

**Comment by cspeterson** [comment]
I'm not sure yet how this project already handles differing output from multiple versions of checkers, so this isn't a PR yet...

but heads up this change works with the newer validate output. I'm using it with Puppet 6.

```diff
diff --git a/ale_linters/puppet/puppet.vim b/ale_linters/puppet/puppet.vim
index ae648615..e6053321 100644
--- a/ale_linters/puppet/puppet.vim
+++ b/ale_linters/puppet/puppet.vim
@@ -8,7 +8,7 @@ function! ale_linters#puppet#puppet#Handle(buffer, lines) abort
-    let l:pattern = '^Error: .*: \(.\+\) \((file:\|at\) .\+\.pp\(, line: \|:\)\(\d\+\)\(, column: \|:\)\=\(\d*\)'
+    let l:pattern = '^Error: \(.\+\) \((file:\|at\) .\+\.pp\(, line: \|:\)\(\d\+\)\(, column: \|:\)\=\(\d*\)'
 
     for l:match in ale#util#GetMatches(a:lines, l:pattern)

```
I've noticed that it is also possible to specify JSON output from `validate` all the back to 3.8 (the earliest documented version), which may well be consistent across versions. I'm seeing some code from others linters that parse through JSON, so I'll likely submit a PR doing that instead of the quick fix above 🙂

**Comment by cspeterson** [comment]
Some updates:

1. The JSON output from `validate` is unhelpful - the JSON is only a summary report, and we still require the line-specific results.
1. The oldest version of puppet actually in any kind of support by Puppetlabs is 5.5.10 according to the following:
  - https://puppet.com/docs/puppet-enterprise/product-support-lifecycle/
  - https://puppet.com/docs/pe/2019.8/component_versions_in_recent_pe_releases.html

And that earliest supported version of Puppet uses the new error reporting format.

Would a PR switching over the ale code completely to the new output format be accepted? Or can someone narrow down what version of puppet changed it?

**Comment by cspeterson** [comment]
A PR for this: #3223
